src/StorageFlattening.cpp

```diff
@@ -400,13 +478,31 @@ class FlattenDimensions : public IRMutator {
     }
 
     Stmt visit(const For *op) override {
+        for (auto &p : hoisted_storages) {
+            Expr expanded_min = simplify(expand_expr(op->min, scope));
+            Expr expanded_extent = simplify(expand_expr(op->extent, scope));
+            Interval loop_bounds = Interval(expanded_min, expanded_min + expanded_extent - 1);
+            p.second.loop_vars.push(op->name, loop_bounds);
+        }
+
         bool old_in_gpu = in_gpu;
         if (op->for_type == ForType::GPUBlock ||
             op->for_type == ForType::GPUThread) {
             in_gpu = true;
         }
         Stmt stmt = IRMutator::visit(op);
         in_gpu = old_in_gpu;
+
+        for (auto &p : hoisted_storages) {
+            p.second.loop_vars.pop(op->name);
+        }
+
+        return stmt;
+    }
+
+    Stmt visit(const LetStmt *op) override {
+        ScopedBinding<Expr> bind(scope, op->name, simplify(expand_expr(op->value, scope)));
```

Comment:
This looks problematic for compile times. Every LetStmt is going to be substituted into its users and then simplified, whether or not it's needed for storage hoisting. It's going to create some combinatorially large expressions.

Reply:
I changed expand_expr to expand expressions recursively, so let will only expanded when it's needed. I think this should be better for compile times (I think SlidingWindow might have the same problem, because that's where I lifted original code from. If you think this approach is better, I can update it there as well).